### ISSUE TITLE
Histogram metric using "quantile" instead of "le"

### ISSUE THREAD

**Issue by christopher-wong** [Ptrskay3/axum-prometheus#24]
A standard histogram looks like this, using the "le" label which causes the `histogram_quantile` function in Grafana to break.

```
# TYPE axum_http_requests_duration_seconds histogram
axum_http_requests_duration_seconds_bucket{method="GET",status="200",endpoint="/metrics",le="0.005"} 2
axum_http_requests_duration_seconds_bucket{method="GET",status="200",endpoint="/metrics",le="0.01"} 2
axum_http_requests_duration_seconds_bucket{method="GET",status="200",endpoint="/metrics",le="0.025"} 2
axum_http_requests_duration_seconds_bucket{method="GET",status="200",endpoint="/metrics",le="0.05"} 2
axum_http_requests_duration_seconds_bucket{method="GET",status="200",endpoint="/metrics",le="0.1"} 2
axum_http_requests_duration_seconds_bucket{method="GET",status="200",endpoint="/metrics",le="0.25"} 2
axum_http_requests_duration_seconds_bucket{method="GET",status="200",endpoint="/metrics",le="0.5"} 2
axum_http_requests_duration_seconds_bucket{method="GET",status="200",endpoint="/metrics",le="1"} 2
axum_http_requests_duration_seconds_bucket{method="GET",status="200",endpoint="/metrics",le="2.5"} 2
axum_http_requests_duration_seconds_bucket{method="GET",status="200",endpoint="/metrics",le="5"} 2
axum_http_requests_duration_seconds_bucket{method="GET",status="200",endpoint="/metrics",le="10"} 2
axum_http_requests_duration_seconds_bucket{method="GET",status="200",endpoint="/metrics",le="+Inf"} 2
axum_http_requests_duration_seconds_sum{method="GET",status="200",endpoint="/metrics"} 0.003824666
axum_http_requests_duration_seconds_count{method="GET",status="200",endpoint="/metrics"} 2
```

A custom metric appears to use "quantile" instead of "le"

```
    metrics::histogram!(
        "lambda_invocation_duration_seconds",
        duration.as_secs_f64(),
        &[("model_id", model_id)]
    );
```

```
# TYPE lambda_invocation_duration_seconds summary
lambda_invocation_duration_seconds{model_id="some_label",quantile="0"} 1.2631185
```

**Comment by Ptrskay3** [comment]
You're free to [setup your own](https://github.com/Ptrskay3/axum-prometheus/blob/master/examples/builder-example/src/main.rs#L32-L41) PrometheusRecorder. Currently I don't think there's a very good reason to change the default in this crate.

**Comment by christopher-wong** [comment]
It's just a little strange that the built in `histogram!` macro records a histogram in a different format than the `axum_http_requests_duration_seconds` histogram, no?

**Comment by christopher-wong** [comment]
Interesting, I'm using the default instantiation so that's surprising.

```rust
let (prometheus_layer, metric_handle) = PrometheusMetricLayer::pair();
...
.layer(prometheus_layer);
```

**Comment by Ptrskay3** [comment]
The point I'm trying to make here is that this crate sets its default `PrometheusHandle` differently from what you'd otherwise use.
https://github.com/Ptrskay3/axum-prometheus/blob/da7630a48107996cd0b6f01a757256957cb91b55/src/lib.rs#L463-L473

That's why I suggested to just override this when you initialize the middleware.

**Comment by yanns** [comment]
Using the `le` label seems to be the default in [Prometheus](https://prometheus.io/docs/practices/histograms/), it'd be great if this crate would use the default label by default.

**Comment by Ptrskay3** [comment]
Sure, I agree. Happy to take PRs. The relevant part is 
https://github.com/Ptrskay3/axum-prometheus/blob/da7630a48107996cd0b6f01a757256957cb91b55/src/lib.rs#L463-L473

**Comment by yanns** [comment]
I've opened https://github.com/metrics-rs/metrics/issues/385

**Comment by yanns** [comment]
For info, the current documentation is using `le`: https://github.com/Ptrskay3/axum-prometheus/blob/da7630a48107996cd0b6f01a757256957cb91b55/src/lib.rs#L74-L85

**Comment by Ptrskay3** [comment]
With my last comment, I specifically meant we should change the default metrics configuration _in this crate_, not in `metrics-exporter-prometheus`, and that's why I highlighted the relevant part to change.

It's great however, that `metrics` authors also confirmed this is in https://github.com/metrics-rs/metrics/issues/385, and suggested us to use `PrometheusBuilder::set_buckets`. The current documentation showing `le` is probably a leftover thing, I'll update as long as we start to use `le` as default eventually.

**Comment by yanns** [comment]
I'm a bit confused.
https://github.com/Ptrskay3/axum-prometheus/blob/da7630a48107996cd0b6f01a757256957cb91b55/src/lib.rs#L463-L473 is already using `set_buckets_for_metric`. Should we add `.set_buckets`? It does not seem to make sense as both are setting the buckets.

**Comment by yanns** [comment]
Maybe the issue is maybe that we are matching on `AXUM_HTTP_REQUESTS_DURATION_SECONDS` that can be changed with `with_prefix`?

**Comment by yanns** [comment]
I can confirm this. If I remove the usage of `with_prefix`, then the metrics are exported with `le`:
```
# TYPE axum_http_requests_duration_seconds histogram
axum_http_requests_duration_seconds_bucket{method="POST",status="200",endpoint="/:project_key/reference-expansion",le="0.005"} 0
axum_http_requests_duration_seconds_bucket{method="POST",status="200",endpoint="/:project_key/reference-expansion",le="0.01"} 0
```

If I use `with_prefix`, the metrics are exported with `quantile`:
```
# TYPE reference_http_requests_duration_seconds summary
reference_http_requests_duration_seconds{method="POST",status="200",endpoint="/:project_key/reference-expansion",quantile="0"} 0.270261
reference_http_requests_duration_seconds{method="POST",status="200",endpoint="/:project_key/reference-expansion",quantile="0.5"} 1.6350935120678547
```

**Comment by yanns** [comment]
A not-ideal fix: https://github.com/Ptrskay3/axum-prometheus/pull/26

**Comment by Ptrskay3** [comment]
I'm doing a larger refactor to make Prometheus optional and to enable other exporters to be used. I think I solved this problem [here](https://github.com/Ptrskay3/axum-prometheus/pull/28/commits/c682c5dd354bc78bc0b7fb2fec27f32e531c4017), by immediately calling `set_prefix` before transitioning into the `Paired` state, so it'll be available for the metric builders.

Could you test #28?

**Comment by Ptrskay3** [comment]
Yeah, I finally see the problem. :sweat_smile:  Fixed it [here](https://github.com/Ptrskay3/axum-prometheus/pull/28/commits/04fc0ca6ab5423bba5ea8db93579f4632531831f), you may try out #28.

**Comment by yanns** [comment]
https://github.com/Ptrskay3/axum-prometheus/pull/28/commits/04fc0ca6ab5423bba5ea8db93579f4632531831f does not fix the issue, sorry. We'll get it! :)

**Comment by Ptrskay3** [comment]
Well, it surely does for me.
![image](https://github.com/Ptrskay3/axum-prometheus/assets/52788117/b7ab3c39-e284-446d-be77-9791c1296471)

**Comment by yanns** [comment]
I'll try again.
But checking:
- I'm using `axum-prometheus = { git = "https://github.com/Ptrskay3/axum-prometheus", commit = "04fc0ca6ab5423bba5ea8db93579f4632531831f" }`
- but I can see the following in the lock file:
```
[[package]]
name = "axum-prometheus"
version = "0.3.4"
source = "git+https://github.com/Ptrskay3/axum-prometheus#b91b74d2d2f16868d8ff9fba8f83ba99b5c4a780"
```

strange.

**Comment by yanns** [comment]
sorry, I was using the wrong syntax...
With the following, it's working:
```
axum-prometheus = { git = "https://github.com/Ptrskay3/axum-prometheus", hash = "04fc0ca6ab5423bba5ea8db93579f4632531831f" }
```

**Comment by Ptrskay3** [comment]
Fixed in #28. Also, [`0.4.0`](https://github.com/Ptrskay3/axum-prometheus/releases/tag/0.4.0) is just released, that includes the fix.
